src/System.Net.NameResolution/src/System/Net/DNS.cs

```diff
@@ -0,0 +1,395 @@
+                asyncResult,
+                CancellationToken.None,
+                TaskCreationOptions.DenyChildAttach,
+                TaskScheduler.Default);
+
+            // Finish the flowing, maybe it completed?  This does nothing if we didn't initiate the flowing above.
+            asyncResult.FinishPostingAsyncOp();
+            return asyncResult;
+        }
+
+        private static IAsyncResult HostResolutionBeginHelper(IPAddress address, bool flowContext, bool includeIPv6, AsyncCallback requestCallback, object state)
+        {
+            if (address == null)
+            {
+                throw new ArgumentNullException("address");
+            }
+
+            if (address.Equals(IPAddress.Any) || address.Equals(IPAddress.IPv6Any))
+            {
+                throw new ArgumentException(SR.net_invalid_ip_addr, "address");
+            }
+
+            GlobalLog.Print("Dns.HostResolutionBeginHelper: " + address);
+
+            // Set up the context, possibly flow.
+            ResolveAsyncResult asyncResult = new ResolveAsyncResult(address, null, includeIPv6, state, requestCallback);
+            if (flowContext)
+            {
+                asyncResult.StartPostingAsyncOp(false);
+            }
+
+            // Start the resolve.
+            Task.Factory.StartNew(
+                s => ResolveCallback(s), 
+                asyncResult,
+                CancellationToken.None,
+                TaskCreationOptions.DenyChildAttach,
+                TaskScheduler.Default);
+
+            // Finish the flowing, maybe it completed?  This does nothing if we didn't initiate the flowing above.
+            asyncResult.FinishPostingAsyncOp();
+            return asyncResult;
+        }
+
+        private static IPHostEntry HostResolutionEndHelper(IAsyncResult asyncResult)
+        {
+            //
+            // parameter validation
+            //
+            if (asyncResult == null)
+            {
+                throw new ArgumentNullException("asyncResult");
+            }
+            ResolveAsyncResult castedResult = asyncResult as ResolveAsyncResult;
+            if (castedResult == null)
+            {
+                throw new ArgumentException(SR.net_io_invalidasyncresult, "asyncResult");
+            }
+            if (castedResult.EndCalled)
+            {
+                throw new InvalidOperationException(SR.Format(SR.net_io_invalidendcall, "EndResolve"));
+            }
+
+            GlobalLog.Print("Dns.HostResolutionEndHelper");
+
+            castedResult.InternalWaitForCompletion();
+            castedResult.EndCalled = true;
+
+            Exception exception = castedResult.Result as Exception;
+            if (exception != null)
+            {
+                throw exception;
+            }
+
+            return (IPHostEntry)castedResult.Result;
+        }
+
+        public static IAsyncResult BeginGetHostEntry(string hostNameOrAddress, AsyncCallback requestCallback, object stateObject)
```

Comment:
Are these Begin/End methods public in the contract?  If not, consider making them internal or private as needed.  (Same goes for any other methods that may not be surfaced.)